selfdrive/statsd.py

```diff
@@ -0,0 +1,104 @@
+#!/usr/bin/env python3
+import os
+import zmq
+import json
+import time
+from pathlib import Path
+from collections import deque
+from datetime import datetime
+from selfdrive.hardware import PC
+from selfdrive.swaglog import cloudlog
+from cereal.messaging import SubMaster
+from common.file_helpers import atomic_write_in_dir
+
+FLUSH_TIME_S = 60
+STATS_SOCKET = "ipc:///tmp/stats"
+if PC:
+  STATS_DIR = os.path.join(str(Path.home()), ".comma", "stats")
+else:
+  STATS_DIR = "/data/stats/"
+
+class METRIC_TYPE:
+  GAUGE = 'g'
+
+class StatLog:
+  def __init__(self):
+    self.pid = None
+
+  def connect(self):
+    self.zctx = zmq.Context()
+    self.sock = self.zctx.socket(zmq.PUSH)
+    self.sock.setsockopt(zmq.LINGER, 10)
+    self.sock.connect(STATS_SOCKET)
+    self.pid = os.getpid()
+
+  def _send(self, metric: str):
+    if os.getpid() != self.pid:
+      self.connect()
+
+    try:
+      self.sock.send_string(metric, zmq.NOBLOCK)
+    except zmq.error.Again:
+      # drop :/
+      pass
+
+  def gauge(self, name: str, value: float):
```

Comment:
not sure if this would be that useful to have per metric, any use cases you can think of? would be easy to add once necessary I imagine

Reply:
tags is also not part of the default statsd protocol. I assumed we wanted to follow that as much a possible.

Reply:
tags will definitely be extremely useful for things like the version, but I guess maybe they will all be static

the alternative is always using metric names namespaced by the tags in the name (e.g. metric `cpu_temp` vs `cpu_temp.0_8_12`) but this can be a lot more difficult to work with

Reply:
oh, things like version etc will all be in the tags by default (see where it generates the lines). I do maybe want to add optional tags to metrics as well where wanted (see adeeb's comment about engaged status for example)